gno/p/daokit/daokit.gno

```diff
@@ -0,0 +1,180 @@
+package daokit
+
+import (
+	"path"
+	"std"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/json"
+	"gno.land/p/demo/mux"
+	"gno.land/p/teritori/daocond"
+)
+
+type Dao struct {
+	Name        string
+	Description string
+
+	MemberModule     *MemberModule
+	ResourcesModule  *ResourcesModule
+	ProposalModule   *ProposalModule
+	MessagesRegistry *MessagesRegistry
+
+	renderingRouter *mux.Router
+	realmName       string
+}
+
+func NewDaoJSON(name, description string, roles []string, members [][]string, resourcesJSON string, handlers []MessageHandler) *Dao {
+	if len(name) < 1 {
+		panic("dao name should be at least 1 characters long")
+	}
+
+	if len(description) < 1 {
+		panic("dao description should be at least 1 characters long")
+	}
+
+	dao := &Dao{
+		Name:             name,
+		Description:      description,
+		MemberModule:     newMemberModule(),
+		ResourcesModule:  newResourcesModule(),
+		ProposalModule:   newProposalModule(),
+		MessagesRegistry: newMessagesRegistry(),
+		renderingRouter:  mux.NewRouter(),
+		realmName:        path.Base(std.CurrentRealm().PkgPath()),
+	}
+	dao.initMessagesRegistry()
+	dao.initRenderingRouter()
+
+	dao.MemberModule.setRoles(roles)
+	dao.MemberModule.setMembers(members)
+
+	resources := dao.parseResourcesJSON(resourcesJSON)
+	conditionRaw, ok := resources.Get("init")
+	if !ok {
+		panic("resources should have an 'init' condition")
+	}
+	condition := conditionRaw.(daocond.Condition)
+	dao.MessagesRegistry.handlers.Iterate("", "", func(key string, value interface{}) bool {
+		resources.Set(key, condition)
+		return false
+	})
+	resources.Remove("init")
+
+	dao.ResourcesModule.setResources(resources)
+
+	for _, handler := range handlers {
+		dao.MessagesRegistry.register(handler)
+	}
+
+	return dao
+}
+
+func (d *Dao) Vote(proposalID uint64, vote string) {
+	voter := std.PrevRealm().Addr()
+	if !d.MemberModule.IsMember(voter.String()) {
+		panic("voter is not a member")
+	}
+
+	proposal := d.ProposalModule.getProposal(proposalID)
+	if proposal == nil {
+		panic("proposal not found")
+	}
+
+	if proposal.status != ProposalStatusOpen {
+		panic("proposal is not open")
+	}
+
+	if daocond.Vote(vote) != daocond.VoteYes && daocond.Vote(vote) != daocond.VoteNo && daocond.Vote(vote) != daocond.VoteAbstain {
+		panic("invalid vote")
+	}
+
+	e := &daocond.EventVote{
+		VoterID: voter.String(),
+		Vote:    daocond.Vote(vote),
+	}
+
+	proposal.state.HandleEvent(e, proposal.votes)
+	proposal.votes[voter.String()] = daocond.Vote(vote)
+
+}
+
+func (d *Dao) Execute(proposalID uint64) {
+	executor := std.PrevRealm().Addr()
+	if !d.MemberModule.IsMember(executor.String()) {
+		panic("executor is not a member")
+	}
+
+	proposal := d.ProposalModule.getProposal(proposalID)
+	if proposal == nil {
+		panic("proposal not found")
+	}
+
+	if proposal.status != ProposalStatusOpen {
```

Comment:
Something to discuss during our point:
State and status seems like too close one to another maybe we need to rename one or another

Reply:
condition state